### ISSUE TITLE
[clang-static-analyzer]:how to promote array or pointer accuracy in clang static analyzer

### ISSUE THREAD

**Issue by mzyKi** [llvm/llvm-project#62678]
```c
// test.c
#define NULL (void*)0
void func(void) {
  char *array = (char *)malloc(10 * sizeof(char));
  if (NULL == array) {
    return;
  }
  array = 'abc';
  clang_analyzer_dump(array[0]);
}
```
run commands on Bash like the following:
```console
$ clang-tidy -checks=-*,clang-analyzer-unix.Malloc,clang-analyzer-debug.ExprInspection test.c --
6 warnings generated.
test.c:8:3: warning: Unknown [clang-analyzer-debug.ExprInspection]
  clang_analyzer_dump(array[0]);
  ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~
test.c:4:7: note: Assuming 'array' is not equal to NULL
  if (NULL == array) {
      ^~~~~~~~~~~~~
test.c:1:14: note: expanded from macro 'NULL'
#define NULL (void*)0
             ^
test.c:4:3: note: Taking false branch
  if (NULL == array) {
  ^
test.c:8:3: note: Unknown
  clang_analyzer_dump(array[0]);

```

here it only can dump `Unknown`

now I want to let it can dump `97 32b` which is `array[0]` ASCII code

how can I make it possible? Thanks for giving me tips.

**Comment by steakhal** [comment]
I didn't quite get the question. It works for me: https://godbolt.org/z/4a7hKqor1
```
void func(void) {
  char *array = "abc";
  clang_analyzer_dump(array[0]); // 97 is reported
}
```

**Comment by mzyKi** [comment]
> I didn't quite get the question. It works for me: https://godbolt.org/z/4a7hKqor1
> 
> ```
> void func(void) {
>   char *array = "abc";
>   clang_analyzer_dump(array[0]); // 97 is reported
> }
> ```

Thanks a lot,my testcase has error so I can not get correct result
